feat: power of product of two reflections

trivial1711 opened the pull request:
Let $M$ be a module over a commutative ring $R$. Let $x, y \in M$ and $f, g \in M^*$ with $f(x) = g(y) = 2$. The corresponding reflections $r_1, r_2 \colon M \to M$ are given by $r_1z = z - f(z) x$ and $r_2 z = z - g(z) y$. These are linear automorphisms of $M$.

We prove that for all $z \in M$ and all $m \in \mathbb{Z}$, we have:
$$(r_1 r_2)^m z = z + S_{\left\lfloor \frac{m-2}{2} \right \rfloor}(t) (S_{\left\lfloor \frac{m-1}{2} \right \rfloor}(t) + S_{\left\lfloor \frac{m-3}{2} \right \rfloor}(t)) (g(x) f(z) y - g(z) y - f(z) x) + S_{\left\lfloor \frac{m-1}{2} \right \rfloor}(t) (S_{\left\lfloor \frac{m}{2} \right \rfloor}(t) + S_{\left\lfloor \frac{m-2}{2} \right \rfloor}(t)) (f(y) g(z) x - f(z) x - g(z) y)$$
$$(r_1 r_2)^m x = (S_{m}(t) + S_{m - 1}(t)) x - S_{m-1} g(x) y$$
$$r_2(r_1 r_2)^m x = (S_{m}(t) + S_{m - 1}(t)) x - S_{m} g(x) y$$
where $t = f(y) g(x) - 2$ and $S_n$ refers to a Chebyshev $S$-polynomial.

These formulas may look ridiculous, but they are necessary for constructing reflection representations of a Coxeter group over an arbitrary ring.

## Appendix: What's coming next
I will now elaborate on this last sentence. You may skip reading this part if you are not interested. Let $W$ be a Coxeter group with simple generators $s_i$ and Coxeter matrix $(M_{i, i'})\_{i, i'}$.

### The traditional way to construct geometric representations
Traditionally, a geometric representation of $W$ is defined in the following way. (For example, see https://www.math.cuhk.edu.hk/course_builder/2223/math6032/lecture-notes-coxeter.pdf, page 9.) Fix any matrix $(k_{i, i})\_{i, i'}$ satisfying the following conditions for all $i, i'$:
1. $k_{i, i} = 2$
2. $k_{i, i'} = 0$ if and only if $M_{i, i'} = 2$
3. $k_{i, i'} \leq 0$ for $i \neq i'$
4. $k_{i, i'} k_{i', i} = 4 \cos^2 (\pi / M_{i, i'})$ if $M_{i, i'} < \infty$
5. $k_{i, i'} k_{i', i} \geq 4$ if $M_{i, i'} = \infty$.

Let $V$ be the free $\mathbb{R}$-vector space on $\alpha_{i}$, and define the element $\alpha_i^\vee \in V^*$ by $\alpha_i^\vee (\alpha\_{i'}) = k_{i, i'}$. Then, one may construct a representation $\rho \colon W \to GL(V)$ which satisfies
$$\rho(s_i) v = v - \alpha_i^\vee(v) \alpha_i$$ for all $i$ and all $v \in V$. This construction requires a rather involved linear algebra computation. It is not unlike the computation in this PR, but it involves a lot of sines and cosines and complex exponentials rather than Chebyshev polynomials.

There are several downsides to the traditional approach that make me think that it is not suitable for mathlib.

#### It only works over $\mathbb{R}$
Suppose that all of the entries of the Coxeter matrix $M$ are in the set $\\{1, 2, 3, 4, 6, \infty\\}$. Then, the numbers $4 \cos^2 (\pi / M_{i, i'})$ appearing in condition $4$ are all integers, and it is possible to do the entire construction over $\mathbb{Z}$ instead of $\mathbb{R}$. This yields a so-called *crystallographic* geometric representation of $W$. I believe that geometric representations of a Coxeter group should be constructed in a way that simply works over $\mathbb{Z}$ (and other ordered rings, such as real cyclotomic fields) to begin with.

#### The vector space is fixed
The traditional approach can be used to construct "the" geometric representation of $W$ (corresponding to a specific generalized Cartan matrix $(k_{i, i})\_{i, i'}$). In "the" geometric representation of $W$, the underlying vector space is "the" free vector space $\bigoplus_i \mathbb{R} \alpha_i$.

This is not great. For example, let $W = S_n$ (the symmetric group, not the Chebyshev $S$-polynomial). This is a Coxeter group generated by the simple transpositions $s_i = (i, i+1)$ for $1 \leq i \leq n - 1$. Let $V$ be the subspace of $\mathbb{R}^n$ consisting of all vectors whose components sum to $0$. Then $V$ is a $W$-module in the usual way. It is common in the informal literature to call the representation $\rho \colon W \to GL(V)$ "the" (standard) geometric representation of $W$, but it isn't. The geometric representation of $W$ has underlying vector space $\bigoplus_i \mathbb{R} \alpha_i = \mathbb{R}^{n-1}$, not $V$. Of course, the two representations are isomorphic, but they are not equal.

In order to get around this issue, we ought to define not only "the" geometric representation of $W$, but also "a" geometric representation of $W$. (See https://arxiv.org/pdf/2405.10387.)

#### It's hard to formalize
The linear algebra computations involved in constructing geometric representations of $W$ are, simply put, quite involved. See the proof of Lemma 20 of https://www.math.cuhk.edu.hk/course_builder/2223/math6032/lecture-notes-coxeter.pdf, which involves computing complex eigenvalues of a 2 x 2 matrix, diagonalizing it, and several other things that would be a nightmare to formalize.

### What should we do instead?
Let $\rho \colon W \to GL(V)$ be any representation, where $V$ is any module over a commutative ring. Let us say that $\rho$ is a *reflection representation* if it sends every simple reflection of $W$ to a reflection in $GL(V)$. That means that there exist elements $\alpha_i \in V$ and $\alpha_i^\vee \in V^*$ such that for all $i$, we have $\alpha_i^\vee (\alpha_i) = 2$, and for all $i$ and all $v \in V$, we have
$$\rho(s_i) v = v - \alpha_i^\vee(v) \alpha_i.$$

How does one construct a reflection representation? By the results of this PR, it is possible to choose the vectors $\alpha_i \in V$ and $\alpha_i^\vee \in V^*$ to be anything, as long as
$$S_{\left\lfloor \frac{m-2}{2} \right \rfloor}(t) (S_{\left\lfloor \frac{m-1}{2} \right \rfloor}(t) + S_{\left\lfloor \frac{m-3}{2} \right \rfloor}(t)) = S_{\left\lfloor \frac{m-1}{2} \right \rfloor}(t) (S_{\left\lfloor \frac{m}{2} \right \rfloor}(t) + S_{\left\lfloor \frac{m-2}{2} \right \rfloor}(t)) = 0 ,$$
for all $i, i'$, where $t = \alpha_i^\vee(\alpha_{i'}) \alpha_{i'}^\vee(\alpha_{i}) - 2$ and $m = M_{i, i'}$. Over the real numbers, it suffices that $\alpha_i^\vee(\alpha_{i'}) \alpha_{i'}^\vee(\alpha_{i}) = 4 \cos^2(j \pi / m)$, where $0 < j < m$.

Suppose that the ring $R$ is ordered. Let us say that $\rho$ is a *geometric representation* if it is a reflection representation, the $\alpha_i$ are linearly independent, and for all $i, i'$ and all integers $k$ with $0 \leq j < M_{i, i'}$, we have that 
$$\rho(\underbrace{\cdots s_i s_{i'} s_i s_{i'}}\_{j})\alpha_i$$
is a nonnegative linear combination of $\alpha_i$ and $\alpha_{i'}$. Here, there are $j$ terms in the product $\cdots s_i s_{i'} s_i s_{i'}$.

How does one construct a geometric representation? By the results of this PR, it suffices that for all $i, i'$, we have $\alpha_i^\vee(\alpha_{i'}) \leq 0$, and $S_j(t) \geq 0$ for $j \leq M_{i, i'} / 2$, where again $t = \alpha_i^\vee(\alpha_{i'}) \alpha_{i'}^\vee(\alpha_{i}) - 2$. Over the real numbers, this is equivalent to the condition that the numbers $k_{i, i'} = \alpha_i^\vee(\alpha_{i'})$ satisfy the conditions 1–5 above.

With these definitions and lemmas, it should still be possible to construct geometric representations in the traditional way, but there will be a more general notion of "a" geometric representation, and the construction can be done over any ring.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [ ] depends on: #13195
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jcommelin commented:
Would you mind copying your appendix to a github issue, and then linking to that issue from here?
I think that is a better location for this explanation, because then it can be linked to from multiple locations.

trivial1711 commented:
> Would you mind copying your appendix to a github issue, and then linking to that issue from here? I think that is a better location for this explanation, because then it can be linked to from multiple locations.

Thanks for the suggestion. I have done it: #13291

jcommelin commented on Mathlib/LinearAlgebra/Reflection.lean:
```diff
@@ -114,6 +121,173 @@ lemma bijOn_reflection_of_mapsTo {Φ : Set M} (h : f x = 2) (h' : MapsTo (reflec
+
+variable {x y : M} {f g : Dual R M} (hf : f x = 2) (hg : g y = 2)
+
+local notation:max "t" => f y * g x - 2
+local notation:max "S_eval_t" n:max => Polynomial.eval t (S R n)
```
I understand that these notations make it easier to work in this file. But outside this file they don't apply. Which means that many of the statements will become rather complicated, and not so easy to handle.

I am not completely sure what the best strategy is. But I think `S_eval_t` could maybe be a standalone definition. (But I'm not sure what the best name is for it.)

jcommelin replied:
What do you think of the following approach? Here I just add `t` as an extra assumption, together with `ht` giving its defining formula:
```
lemma reflection_mul_reflection_pow_apply (m : ℕ) (z : M) (t : R) (ht : t = f y * g x - 2) :
    ((reflection hf * reflection hg) ^ m) z =
      z +
        ((S R ((m - 2) / 2)).eval t * ((S R ((m - 1) / 2)).eval t + (S R ((m - 3) / 2)).eval t)) •
          ((g x * f z - g z) • y - f z • x) +
        ((S R ((m - 1) / 2)).eval t * ((S R (m / 2)).eval t + (S R ((m - 2) / 2)).eval t)) •
          ((f y * g z - f z) • x - g z • y) := by
```

jcommelin replied:
If we can convince the pretty printer to use dot-notation for `Polynomial.eval` (see [zulip](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Polynomial.2Eeval.20dot-notation.20delaborator)), then I think this is quite readable, and doesn't need local notation.

jcommelin commented on Mathlib/LinearAlgebra/Reflection.lean:
```diff
@@ -114,6 +121,173 @@ lemma bijOn_reflection_of_mapsTo {Φ : Set M} (h : f x = 2) (h' : MapsTo (reflec
+    -- Expand out all the reflections and use `hf`, `hg`.
+    simp only [reflection_apply, map_add, map_sub, map_smul, hf, hg]
+    -- `m` can either be written in the form `2 * k` or `2 * k + 1`.
+    obtain ⟨k, even | odd⟩ := Int.even_or_odd' m
+    · /- The goal contains many expressions of the form `S_eval_t n`. We may rewrite all of them as
+      `S_eval_t (-2 + k)`, `S_eval_t (-1 + k)`, or `S_eval_t k`. -/
+      push_cast
+      simp_rw [even, add_sub_assoc (2 * k), sub_eq_add_neg (2 * k), add_comm (2 * k),
+        add_mul_ediv_left _ k (by norm_num : (2 : ℤ) ≠ 0)]
+      norm_num
+      /- Now, equate the coefficients on both sides. These linear combinations were
+      found using `polyrith`. -/
+      match_scalars
+      · rfl
+      · linear_combination (norm := ring_nf) (-g z * f y * S_eval_t (-1 + k) +
+            f z * S_eval_t (-1 + k)) * S_eval_t_sub_two k +
+          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+      · linear_combination (norm := ring_nf) g z * S_eval_t (-1 + k) * S_eval_t_sub_two k +
+          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+    · /- The goal contains many expressions of the form `S_eval_t n`. We may rewrite all of them as
+      `S_eval_t (-1 + k)`, `S_eval_t k`, or `S_eval_t (1 + k)`. -/
+      push_cast
+      simp_rw [odd, add_assoc (2 * k), add_sub_assoc (2 * k), add_comm (2 * k),
+        add_mul_ediv_left _ k (by norm_num : (2 : ℤ) ≠ 0)]
+      norm_num
+      /- Now, equate the coefficients on both sides. These linear combinations were
+      found using `polyrith`. -/
+      match_scalars
+      · rfl
+      · linear_combination (norm := ring_nf) (-g z * f y * S_eval_t k +
+            f z * S_eval_t k) * S_eval_t_sub_two (k + 1) +
+          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+      · linear_combination (norm := ring_nf) g z * S_eval_t k * S_eval_t_sub_two (k + 1) +
+          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
```
What do you think of the following way to finish the proof?
```suggestion
    push_cast
    set k : ℤ := m / 2
    set e : ℤ := m % 2
    rw [show m = 2 * k + e from (Int.ediv_add_emod m 2).symm]
    simp_rw [add_assoc (2 * k), add_sub_assoc (2 * k), add_comm (2 * k),
      add_mul_ediv_left _ k (by norm_num : (2 : ℤ) ≠ 0)]
    have he : e = 0 ∨ e = 1 := by omega
    clear_value e
    set t' := t; subst ht; set t := t'
    match_scalars
    · rfl
    · linear_combination (norm := skip) (-g z * f y * (S R (e - 1 + k)).eval t +
          f z * (S R (e - 1 + k)).eval t) * S_eval_t_sub_two (e + k) +
          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
      obtain rfl | rfl : e = 0 ∨ e = 1 := he <;> ring_nf
    · linear_combination (norm := skip)
          g z * (S R (e - 1 + k)).eval t * S_eval_t_sub_two (e + k) +
          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
      obtain rfl | rfl : e = 0 ∨ e = 1 := he <;> ring_nf
```

jcommelin commented on Mathlib/LinearAlgebra/Reflection.lean:
```diff
@@ -114,6 +121,173 @@ lemma bijOn_reflection_of_mapsTo {Φ : Set M} (h : f x = 2) (h' : MapsTo (reflec
+          (-g z * f y + f z) * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+      · linear_combination (norm := ring_nf) g z * S_eval_t k * S_eval_t_sub_two (k + 1) +
+          g z * S_eval_t_sq_add_S_eval_t_sq (k - 1)
+
```
It might be useful to have an unapplied version of the lemma:
```
/-- A formula for $(r_1 r_2)^m$, where $m$ is a natural number. -/
lemma reflection_mul_reflection_pow' (m : ℕ) (t : R) (ht : t = f y * g x - 2) :
    ((reflection hf * reflection hg) ^ m) =
      LinearMap.id (R := R) (M := M) +
        (((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2))).eval t) •
          ((g x • f - g).smulRight y - f.smulRight x)) +
        (((S R ((m - 1) / 2) * (S R (m / 2) + S R ((m - 2) / 2))).eval t) •
          ((f y • g - f).smulRight x - g.smulRight y)) := by
  ext z
  simp
  rw [← reflection_mul_reflection_pow_apply hf hg m z t ht]
  rw [LinearMap.pow_apply, LinearEquiv.pow_apply]
  rfl
```

trivial1711 commented:
Thank you so much for the feedback.

I added the hypotheses `t` and `ht` to all the lemmas in this section, as you suggested. I also gave them default values: `(t : R := f y * g x - 2) (ht : t = f y * g x - 2 := by rfl)`.

I added `reflection_mul_reflection_pow` and `reflection_mul_reflection_zpow`, which are unapplied versions of `reflection_mul_reflection_pow_apply` and `reflection_mul_reflection_zpow_apply` respectively. Currently, both `reflection_mul_reflection_pow` and `reflection_mul_reflection_zpow` are about `((reflection hf * reflection hg) ^ m).toLinearMap` instead of `(reflection hf * reflection hg) ^ m`. I could change it so that `reflection_mul_reflection_pow` is about `(reflection hf * reflection hg) ^ m` but `reflection_mul_reflection_zpow` is about `((reflection hf * reflection hg) ^ m).toLinearMap`, if you think that would be better.

Finally, I changed the proof of `reflection_mul_reflection_pow_apply` as you suggested, which greatly simplified it.

trivial1711 commented:
Do you think it is better to state the lemmas using `((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2)).eval t` or `(S R ((m - 2) / 2)).eval t * ((S R ((m - 1) / 2)).eval t + (S R ((m - 3) / 2)).eval t)`? I think I should change all of them to be consistent.

trivial1711 commented:
> Do you think it is better to state the lemmas using `((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2)).eval t` or `(S R ((m - 2) / 2)).eval t * ((S R ((m - 1) / 2)).eval t + (S R ((m - 3) / 2)).eval t)`? I think I should change all of them to be consistent.

I think I prefer the latter. The reason is that `((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2)).eval t` is not in simp normal form because of [Polynomial.eval_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Polynomial/Eval/Defs.html#Polynomial.eval_add). (It is questionable whether `Polynomial.eval_add` should be a simp lemma in the first place because it has two `x`s on the RHS and only one on the LHS, but as long as it is, `(S R ((m - 2) / 2)).eval t * ((S R ((m - 1) / 2)).eval t + (S R ((m - 3) / 2)).eval t)` should be better than `((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2)).eval t`.)

trivial1711 commented:
But on the other hand, `((S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2)).eval t` is easier to reason with (for example, it is obviously equivalent to the statement that `t` is a root of the polynomial `(S R ((m - 2) / 2) * (S R ((m - 1) / 2) + S R ((m - 3) / 2))`)

Should I make a separate pull request that removes the @[simp] attribute from Polynomial.eval_mul and Polynomial.eval_add?

jcommelin commented:
Let's just stick to the current normal forms, and I agree it would then be better to have the version with 3 `eval t`s.

The fact that `Polynomial.eval_{mul,add}` is a simp lemma is consistent with the fact that for any monoid hom we have `f (x * y) =  f x * f y` as a simp lemma. So I don't think we should change that.

jcommelin reviewed (commented):
I'll leave it up to you to decide on whether you want to put the algebra inside or outside `.eval t`.
The rest LGTM.

Thanks! :tada: 

bors d+

trivial1711 commented:
bors r+
